src/lock_free/mwcas_descriptor.cpp

```diff
@@ -0,0 +1,206 @@
+{
+namespace
+{
+/*##############################################################################
+ * Local constants
+ *############################################################################*/
+
+/// @brief An offset for right-shifting to extract the "begin position".
+constexpr uint64_t kPosShift = 47;
+
+/// @brief An offset for right-shifting to extract the "reference counter".
+constexpr uint64_t kCntShift = 50;
+
+/// @brief A constant for incrementing the "reference counter".
+constexpr uint64_t kCntUnit = 1UL << kCntShift;
+
+/// @brief A bitmask with only the "begin position" portion set to 1.
+constexpr uint64_t kPosMask = (((1UL << (1 + kCntShift - kPosShift)) - 1) << kPosShift);
+
+/// @brief A bitmask with only the "reference counter" portion set to 1.
+constexpr uint64_t kCntMask = (((1UL << (64 - kCntShift)) - 1) << kCntShift);
+
+/// @brief A bitmask with only the "MwCAS FLAG" and "descriptor address" portions set to 1.
+constexpr uint64_t kDescMask = kMwCASFlag | MwCASDescriptor::kAddrMask;
+
+/*##############################################################################
+ * Local global variable
+ *############################################################################*/
+thread_local std::unique_ptr<MwCASDescriptor> tls = nullptr;
+
+}  // namespace
+
+/*##############################################################################
+ * Static utilities
+ *############################################################################*/
+
+auto
+MwCASDescriptor::GetDescriptor()  //
+    -> MwCASDescriptor *
+{
+  auto *desc = tls.release();
+  if (!desc) {
+    desc = ::dbgroup::memory::Allocate<MwCASDescriptor>();
+  }
+  desc->target_cnt_ = 0;
+  desc->exit_cnt_.store(0, kRelaxed);
+  return desc;
+}
+
+/*##############################################################################
+ * Utilities
+ *############################################################################*/
+auto
+MwCASDescriptor::MwCAS()  //
+    -> bool
+{
+  // set a memory fence
+  stat_.store(kUndecided, kRelease);
+  const auto succeeded = MwCASInternal();
+  return succeeded;
+}
+
+auto
+MwCASDescriptor::MwCASInternal(  //
+    const size_t begin_pos       //
+    ) -> bool
+{
+  const auto desc_addr = std::bit_cast<uint64_t>(this) | kMwCASFlag;
+  auto cur_stat = stat_.load(kSeqCst);
+  if (cur_stat == kUndecided) {
+    auto stat = kSucceeded;
+    for (size_t i = begin_pos; i < target_cnt_; ++i) {
+      if (!EmbedDescriptor(desc_addr, i)) {
+        stat = kFailed;
+        break;
+      }
+    }
+    cur_stat = stat_.load(kSeqCst);
+    if (cur_stat == kUndecided && stat_.compare_exchange_strong(cur_stat, stat, kSeqCst, kSeqCst)) {
+      cur_stat = stat;
+    }
+  }
+
+  const auto succeeded = (cur_stat == kSucceeded);
+  auto follow_cnt = 0;
+  if (succeeded) {
+    for (size_t i = 0; i < target_cnt_; ++i) {
+      auto &target = targets_[i];
+      UnembedDescriptor(desc_addr, target, &follow_cnt, target.new_val);
```

Comment:
呼び出し方としてはこんな感じです．なお， `follow_cnt` が負の値になることはないので `uint32_t` でよいです（定数の後に `U` をつけると符号なしに， `L` をつけると64bit扱いになります（例： `0` なら `int32_t` ， `0UL` なら `uint64_t`））．

```suggestion
  auto follow_cnt = 0U;
  if (succeeded) {
    for (size_t i = 0; i < target_cnt_; ++i) {
      auto &target = targets_[i];
      follow_cnt += Finalize(desc_addr, target, target.new_val);
```